projection/chainstats/chainstats.go

```diff
@@ -0,0 +1,158 @@
+	"strconv"
+
+	"github.com/crypto-com/chain-indexing/internal/utctime"
+	"github.com/pkg/errors"
+
+	"github.com/crypto-com/chain-indexing/appinterface/projection/rdbprojectionbase"
+	"github.com/crypto-com/chain-indexing/appinterface/rdb"
+	event_entity "github.com/crypto-com/chain-indexing/entity/event"
+	entity_projection "github.com/crypto-com/chain-indexing/entity/projection"
+	applogger "github.com/crypto-com/chain-indexing/internal/logger"
+	"github.com/crypto-com/chain-indexing/projection/chainstats/view"
+	event_usecase "github.com/crypto-com/chain-indexing/usecase/event"
+)
+
+var _ entity_projection.Projection = &ChainStats{}
+
+const GENESIS_BLOCK_TIME = "genesis_block_time"
+const TOTAL_BLOCK_TIME = "total_block_time"
+const TOTAL_BLOCK_COUNT = "total_block_count"
+
+type ChainStats struct {
+	*rdbprojectionbase.Base
+
+	rdbConn rdb.Conn
+	logger  applogger.Logger
+
+	maybeGenesisBlockTime *int64 // Unix Nano
+	maybeTotalBlockCount  *big.Int
+}
+
+func NewChainStats(logger applogger.Logger, rdbConn rdb.Conn) *ChainStats {
+	return &ChainStats{
+		rdbprojectionbase.NewRDbBase(rdbConn.ToHandle(), "ChainStats"),
+
+		rdbConn,
+		logger,
+
+		nil,
+		nil,
+	}
+}
+
+func (_ *ChainStats) GetEventsToListen() []string {
+	return []string{
+		event_usecase.GENESIS_CREATED,
+		event_usecase.BLOCK_CREATED,
+	}
+}
+
+func (projection *ChainStats) OnInit() error {
+	chainStatsView := view.NewChainStats(projection.rdbConn.ToHandle())
+
+	var getErr error
+	rawGenesisBlockTime, getErr := chainStatsView.FindBy(GENESIS_BLOCK_TIME)
+	if getErr != nil {
+		return fmt.Errorf("error getting genesis block time on init: %v", getErr)
+	}
+
+	if rawGenesisBlockTime == "" {
+		return nil
+	}
+
+	unixNanoTime, parseErr := strconv.ParseInt(rawGenesisBlockTime, 10, 64)
+	if parseErr != nil {
+		return fmt.Errorf("error parsing genesis block time on init: %v", parseErr)
+	}
+	projection.maybeGenesisBlockTime = &unixNanoTime
+
+	rawTotalBlockCount, getErr := chainStatsView.FindBy(TOTAL_BLOCK_COUNT)
+	if getErr != nil {
+		return fmt.Errorf("error getting total block count on init: %v", getErr)
+	}
+	var ok bool
+	if projection.maybeTotalBlockCount, ok = new(big.Int).SetString(rawTotalBlockCount, 10); !ok {
+		return errors.New("error parsing total block count as big.Int on init")
+	}
+
+	return nil
+}
+
+func (projection *ChainStats) HandleEvents(height int64, events []event_entity.Event) error {
+	rdbTx, err := projection.rdbConn.Begin()
+	if err != nil {
+		return fmt.Errorf("error beginning transaction: %v", err)
+	}
+
+	committed := false
+	defer func() {
+		if !committed {
+			_ = rdbTx.Rollback()
+		}
+	}()
+
+	rdbTxHandle := rdbTx.ToHandle()
+	chainStatsView := view.NewChainStats(rdbTxHandle)
+
+	event := events[0]
+	var maybeGenesisBlockTime *int64
+	if genesisCreatedEvent, ok := event.(*event_usecase.GenesisCreated); ok {
+		genesisBlockTime, parseErr := utctime.Parse(
+			"2006-01-02T15:04:05.000000Z", genesisCreatedEvent.Genesis.GenesisTime,
```

Comment:
Which genesis timestamp it is?
Should it be CROs genesis TS?

Reply:
Not really. Here we do not assume any chain. This time is the layout string as specified in https://golang.org/pkg/time/#Parse